### PR TITLE
[FLUX] Support ControlNet

### PR DESCRIPTION
**Pull request by wangqixun**
# What does this PR do?

1 Support FLUX ControlNet
2 New FluxControlNetPipeline and Demo

The [current weights](https://huggingface.co/InstantX/FLUX.1-dev-controlnet-canny-alpha) are trained on 512x512, but inference can still be performed on non-512 sizes. The latest 1024 + multi-scale model is under training, and it will be synchronized and open-sourced on HF afterwards.

`512x512`
image_demo (1)

`1024x1024` 
image_demo

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@yiyixuxu @haofanwang

### CODE REVIEW

File: src/diffusers/pipelines/flux/pipeline_flux_controlnet.py
Comment (sayakpaul) [inline src/diffusers/pipelines/flux/pipeline_flux_controlnet.py:28]: Should be `FluxLoraLoaderMixin`.

File: src/diffusers/pipelines/flux/pipeline_flux_controlnet.py
Comment (sayakpaul) [inline src/diffusers/pipelines/flux/pipeline_flux_controlnet.py:62]:
```suggestion
        >>> from diffusers import FluxControlNetPipeline
        >>> from diffusers import FluxControlNetModel
```

We should able to do just that:

File: src/diffusers/pipelines/flux/pipeline_flux_controlnet.py
Comment (sayakpaul) [inline src/diffusers/pipelines/flux/pipeline_flux_controlnet.py:91]: Missing "Copied from ..." statement.

File: src/diffusers/models/transformers/transformer_flux.py
Comment (sayakpaul) [inline src/diffusers/models/transformers/transformer_flux.py:458]: Sleek!

File: src/diffusers/pipelines/flux/pipeline_flux_controlnet.py
Comment (sayakpaul) [inline src/diffusers/pipelines/flux/pipeline_flux_controlnet.py:179]: This should be T5. Please follow the latest docstring.

File: src/diffusers/pipelines/flux/pipeline_flux_controlnet.py
Comment (sayakpaul) [inline src/diffusers/pipelines/flux/pipeline_flux_controlnet.py:308]: Missing copy statement.

File: src/diffusers/pipelines/flux/pipeline_flux_controlnet.py
Comment (sayakpaul) [inline src/diffusers/pipelines/flux/pipeline_flux_controlnet.py:393]: This should include a check on images as well? Example:
https://github.com/huggingface/diffusers/blob/8e3affc6690d30a3adbb3bbe0869171ea863ccef/src/diffusers/pipelines/controlnet/pipeline_controlnet.py#L589

File: src/diffusers/pipelines/flux/pipeline_flux_controlnet.py
Comment (sayakpaul) [inline src/diffusers/pipelines/flux/pipeline_flux_controlnet.py:441]: "Copied from" missing.

File: src/diffusers/pipelines/flux/pipeline_flux_controlnet.py
Comment (sayakpaul) [inline src/diffusers/pipelines/flux/pipeline_flux_controlnet.py:442]: "Copied from" missing.

File: src/diffusers/pipelines/flux/pipeline_flux_controlnet.py
Comment (sayakpaul) [inline src/diffusers/pipelines/flux/pipeline_flux_controlnet.py:456]: "Copied from" missing.

File: src/diffusers/pipelines/flux/pipeline_flux_controlnet.py
Comment (sayakpaul) [inline src/diffusers/pipelines/flux/pipeline_flux_controlnet.py:465]: "Copied from" missing.

File: src/diffusers/pipelines/flux/pipeline_flux_controlnet.py
Comment (sayakpaul) [inline src/diffusers/pipelines/flux/pipeline_flux_controlnet.py:478]: "Copied from" missing.

File: src/diffusers/pipelines/flux/pipeline_flux_controlnet.py
Comment (sayakpaul) [inline src/diffusers/pipelines/flux/pipeline_flux_controlnet.py:794]: Do we need this?

**Review by sayakpaul — COMMENTED** [review]
Thanks so much! This is already in an extremely good shape. Just a few comments. Then let's add tests and docs and ship 🚀

**Comment by haofanwang** [comment]
@sayakpaul Formatted. The quality check fails because of other files.

File: src/diffusers/pipelines/flux/pipeline_flux_controlnet.py
Comment (yiyixuxu) [reply src/diffusers/pipelines/flux/pipeline_flux_controlnet.py:440]: can we test out if this works? I think it should be 
```
@staticmethod
# Copied from ...
def _prepare_latent_image_ids
...
```

**Review by yiyixuxu — APPROVED** [review]
looks great!
should add a test before merge

File: src/diffusers/models/controlnet_flux.py
Comment (sayakpaul) [reply src/diffusers/models/controlnet_flux.py:274]: Nie catch thanks. I think the author will group the suggestions together and push :)

Comment (wangqixun) [reply src/diffusers/models/controlnet_flux.py:275]: Added this modification.

**Comment by haofanwang** [comment]
@yiyixuxu @sayakpaul Done. Could you review it?

**Comment by sayakpaul** [comment]
@DN6 this looks quite good to me. Will defer to you for the final approval and merge. 

@linoytsaban @apolinario @asomoza  in case you wanna take it to a spin!

**Comment by innokean** [comment]
Would it be possible to support device_map='balanced' in FluxControlNetModel?
(the model class needs to implement the `_no_split_modules` attribute)

**Comment by sayakpaul** [comment]
@innokean let's do that in a separate PR once this is is shipped?

**Comment by sayakpaul** [comment]
@wangqixun @haofanwang I could not push directly to the branch (I assumed I could because usually maintainers can do that) but the doc example needs fixing:

```py
        >>> import torch
        >>> from diffusers.utils import load_image
        >>> from diffusers import FluxControlNetPipeline
        >>> from diffusers import FluxControlNetModel

        >>> controlnet_model = "InstantX/FLUX.1-dev-controlnet-canny-alpha"
        >>> base_model = "black-forest-labs/FLUX.1-dev"
        >>> controlnet = FluxControlNetModel.from_pretrained(controlnet_model, torch_dtype=torch.bfloat16)
        >>> pipe = FluxControlNetPipeline.from_pretrained(
        ...     base_model, controlnet=controlnet, torch_dtype=torch.bfloat16
        ... )
        >>> pipe.to("cuda")
        >>> control_image = load_image("https://huggingface.co/InstantX/SD3-Controlnet-Canny/resolve/main/canny.jpg")
        >>> prompt = "A girl in city, 25 years old, cool, futuristic"
        >>> image = pipe(
        ...     prompt,
        ...     control_image=control_image,
        ...     controlnet_conditioning_scale=0.6,
        ...     num_inference_steps=28,
        ...     guidance_scale=3.5,
        ... ).images[0]

        >>> image.save("flux.png")
```

**Comment by haofanwang** [comment]
@sayakpaul formatted.

File: src/diffusers/models/controlnet_flux.py
Comment (DN6) [inline src/diffusers/models/controlnet_flux.py:40]: `FromOriginalModelMixin` allows loading the checkpoint in the format orginally shipped by Black Forest Labs. It also requires adding a conversion function to `single_file_utils` and a model mapping to `single_file_model` in the loaders module.   

@wangqixun If you have a checkpoint in this format on hand would you mind sharing a link? We can add support for it in a follow up PR. If not, then we can remove this Mixin for now.

Comment (wangqixun) [reply src/diffusers/models/controlnet_flux.py:40]: I don't have the checkpoint either, and I'm not sure if it would work. It is recommended to remove it first.

Comment (DN6) [reply src/diffusers/models/controlnet_flux.py:40]: Yeah let's remove then please 👍🏽

**Comment by DN6** [comment]
@wangqixun I think you might have to run `make style && make quality` once on the PR. 

We can merge once CI is green 👍🏽

**Comment by haofanwang** [comment]
@DN6 We have run `make style && make quality` locally, but the errors are from other files. Do we need to fix them at this PR?

```
tests/schedulers/test_schedulers.py:865:16: E721 Use `is` and `is not` for type comparisons, or `isinstance()` for isinstance checks
    |
863 |         scheduler_loaded = DDIMScheduler.from_pretrained(self.org_repo_id)
864 | 
865 |         assert type(scheduler) == type(scheduler_loaded)
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ E721
866 | 
867 |         # Reset repo
    |

Found 17 errors.
make: *** [style] Error 1
```

**Comment by DN6** [comment]
@haofanwang What version of `ruff` do you have installed in your environment?

**Comment by haofanwang** [comment]
@DN6 `ruff 0.5.6`

**Comment by DN6** [comment]
@haofanwang Ah diffusers is still pinned to a much older version. Could you install `ruff==0.1.5` and rerun so we can merge? I'll update the ruff version in a separate PR

**Comment by haofanwang** [comment]
@DN6 Done. It works now.

File: src/diffusers/pipelines/flux/pipeline_flux_controlnet.py
Comment (DN6) [inline src/diffusers/pipelines/flux/pipeline_flux_controlnet.py:515]: Think this should be 
```suggestion
    # Copied from diffusers.pipelines.controlnet.pipeline_controlnet.StableDiffusion3ControlNetPipeline.prepare_image
```

File: src/diffusers/pipelines/flux/pipeline_flux_controlnet.py
Comment (DN6) [inline src/diffusers/pipelines/flux/pipeline_flux_controlnet.py:441]:
```suggestion
    @staticmethod
    # Copied from diffusers.pipelines.flux.pipeline_flux.FluxPipeline._unpack_latents
```

File: src/diffusers/pipelines/flux/pipeline_flux_controlnet.py
Comment (DN6) [inline src/diffusers/pipelines/flux/pipeline_flux_controlnet.py:458]:
```suggestion
    @staticmethod
    # Copied from diffusers.pipelines.flux.pipeline_flux.FluxPipeline._pack_latents
```

File: src/diffusers/pipelines/flux/pipeline_flux_controlnet.py
Comment (DN6) [inline src/diffusers/pipelines/flux/pipeline_flux_controlnet.py:467]:
```suggestion
    @staticmethod
    # Copied from diffusers.pipelines.flux.pipeline_flux.FluxPipeline._unpack_latents
```

File: src/diffusers/pipelines/flux/pipeline_flux_controlnet.py
Comment (DN6) [inline src/diffusers/pipelines/flux/pipeline_flux_controlnet.py:442]:
```suggestion
    @staticmethod
    # Copied from diffusers.pipelines.flux.pipeline_flux.FluxPipeline._prepare_latent_image_ids
```

File: src/diffusers/models/controlnet_flux.py
Comment (DN6) [inline src/diffusers/models/controlnet_flux.py:118]:
```suggestion
        def fn_recursive_add_processors(name: str, module: torch.nn.Module, processors: Dict[str, AttentionProcessor]):
```

**Comment by DN6** [comment]
Just left some comments where the Copied from isn't working.

**Comment by haofanwang** [comment]
@DN6 Looks like the [error](https://github.com/huggingface/diffusers/actions/runs/10363869960/job/28688851914?pr=9126) is unrelated?

**Comment by haofanwang** [comment]
@DN6 @sayakpaul Ready to merge.

**Comment by DN6** [comment]
@haofanwang Would it be possible to do a follow up PR with the docs for the Flux ControlNet?

**Comment by Laidawang** [comment]
@wangqixun great job! Can you share your training script?

**Comment by yinian-lw** [comment]
@wangqixun great job! Do it only support Flux 1. dev ? I run on Flux 1. schnell revieved error. [https://github.com/huggingface/diffusers/issues/9309](url)

**Pull request merged**
